File: src/board/BoardV3.go
package board

import (
	"github.com/HuXin0817/dots-and-boxes/src/config"
	"github.com/HuXin0817/dots-and-boxes/src/container"
	"github.com/HuXin0817/dots-and-boxes/src/model"
	"github.com/stretchr/testify/assert"
)

type V3 struct {
	V1
	scoreableEdges container.EdgeQueue
}

func NewV3() *V3 {
	return &V3{
		V1: *NewV1(),
	}
}

func (b *V3) Reset(nb *V1) {
	b.V1 = *nb
	b.scoreableEdges.Clear()
}

func (b *V3) Add(e model.Edge) (s int) {
	s = b.V1.Add(e)
	for _, box := range model.NearBoxes[e] {
		if b.EdgeCountOfBox[box] == 3 {
			b.scoreableEdges.Append(b.findNotContainsEdgeInBox(box))
		}
	}
	return s
}

func (b *V3) MaxObtainableScore(minScore int) (s int) {
	for b.Gaming() {
		if b.scoreableEdges.Empty() {
			if e := b.findScoreableEdge(); e != -1 {
				b.scoreableEdges.Append(e)
			} else {
				break
			}
		}
		e := b.scoreableEdges.Pop()
		if b.Contains(e) {
			continue
		}
		s0 := b.Add(e)
		if config.DEBUG {
			assert.True(nil, s0 > 0)
		}
		s += s0
		if s >= minScore {
			break
		}
	}
	return s
}

File: src/board/BoardV3.cpp
#include "BoardV3.h"

void
BoardV3::Reset(const BoardV1& nb) {
  GetBoardV1() = nb;
  ScoreableEdges.Clear();
}

int
BoardV3::Add(Edge edge) {
  int score = BoardV1::Add(edge);
  for (auto box : EdgeBoxMapper::EdgeNearBoxes[edge]) {
    if (EdgeCountOfBox::operator[](box) == 3) {
      Edge edgeToAdd = FindNotContainsEdgeInBox(box);
      ScoreableEdges.Append(edgeToAdd);
    }
  }
  return score;
}

int
BoardV3::MaxObtainableScore(int minScore) {
  int score = 0;
  while (Gaming()) {
    if (ScoreableEdges.Empty()) {
      if (Edge e = FindScoreableEdge(); static_cast<int>(e) != -1) {
        ScoreableEdges.Append(e);
      } else {
        break;
      }
    }
    Edge edge = ScoreableEdges.Pop();
    if (Contains(edge)) {
      continue;
    }
    int addScore = Add(edge);
    assert(addScore > 0);
    score += addScore;
    if (score >= minScore) {
      break;
    }
  }
  return score;
}

[[nodiscard]] bool
BoardV3::ScoreableEdgesEmpty() const {
  return ScoreableEdges.Empty();
}
